### ISSUE TITLE
Upgrade from 2.0 to 2.5 db migrations canceled

### ISSUE THREAD

**Issue by veritaspoutine** [zammad/zammad#2159]
### Infos:

* Used Zammad version: 2.5
* Installation method (source, package, ..): RPM Update 
* Operating system: CentOS 7
* Database + version: postgres 9.2.23
* Elasticsearch version: 5.6
* Browser + version: chrome, firefox, ie

### Expected behavior:
I am upgrading from 2.0 to 2.5

- DB migration succeed. 
- Sending and fetching mails work flawlessly. 

### Actual behavior:
- rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

Need data_option[:null] param with true or false
/opt/zammad/app/models/object_manager/attribute.rb:900:in `check_datatype'

- Outgoing mail encounter issue: Can't use Channel::Driver::Smtp: #<TypeError: no implicit conversion of nil into String>" 
![image](https://user-images.githubusercontent.com/14975224/43517091-1e3686ae-95ba-11e8-9f55-a1b7e48fdc59.png)
![image](https://user-images.githubusercontent.com/14975224/43517135-41bcd45c-95ba-11e8-8a33-4b9c85a540c0.png)

While researching for possible solutions, I came across #2138 that eventually lead me to #2106. I tried the following suggested steps:

1. systemctl stop zammad
2. zammad run rails r 'Cache.clear'
3. zammad run rake db:migrate
4. systemctl start zammad
5. Reload your browser

But after executing zammad run rake db:migrate, I got the same rake aborted error too!
![image](https://user-images.githubusercontent.com/14975224/43516918-8bedb24a-95b9-11e8-99dd-ce718068ed4c.png)

Not sure if I am supposed to directly upgrade to 2.5 from 2.0. I would be really appreciated if someone can shed some lights. 

edit: 
zammad logs gave me this error message
![image](https://user-images.githubusercontent.com/14975224/43519587-9dc95baa-95c2-11e8-9fe8-4375f349f12a.png)

### Steps to reproduce the behavior:

CentOS 7 rpm upgrade to 2.5 from 2.0. 

edit: 
restoring DB and file from 2.0 backup to fresh install Zammad 2.5 resulted the same issue. 

Yes I'm sure this is a bug and no feature request or a general question.

**Comment by thorsteneckel** [comment]
Hi @veritaspoutine - we actually did multiple changes over the versions which seems to lead to this issue. As a workaround: Would it be possible for you to migrate from version to version like 2.0 -> 2.1 -> 2.2 -> 2.3 -> 2.4 -> 2.5 ?

PS: Thanks for your comprehensive writeup!

**Comment by veritaspoutine** [comment]
hey @thorsteneckel thanks for replying! i don't see 2.1, 2.2, 2.3, and 2.4 in the zammad repo. Could you point me where I could get them?

edit:
I am asking for rpm package because updating to 2.1 using source would produce the following message:
"Your Ruby version is 2.0.0, but your Gemfile specified 2.4.1" 
and this is where I stop before messing up dependencies.

**Comment by thorsteneckel** [comment]
Hi! Sure - we're prefixing the the version releases with `stable-` so it's e.g. `stable-2.1` more information about the package / repo can be found on the corresponding page over at packager.io: http://packager.io/gh/zammad/zammad/refs/stable-2.1

**Comment by thorsteneckel** [comment]
One of our paying customers faced the same issue. The structure of the field causing this issue looks like this:

```
  {
    "id": 1337,
    "object_lookup_id": 1,
    "name": "example_field",
    "display": "Example Field",
    "data_type": "input",
    "data_option": {},
    "data_option_new": {},
    "editable": true,
    "active": true,
    "screens": {
      "create": {
        "ticket.agent": {
          "shown": true
        },
        "admin.user": {
          "shown": true
        }
      },
      "edit": {
        "ticket.agent": {
          "shown": true
        },
        "admin.user": {
          "shown": true
        }
      },
      "view": {
        "ticket.agent": {
          "shown": true
        },
        "admin.user": {
          "shown": true
        }
      }
    },
    "to_create": false,
    "to_migrate": false,
    "to_delete": false,
    "to_config": false,
    "position": 1550,
    "created_by_id": 8,
    "updated_by_id": 8,
    "created_at": "2018-01-21T11:51:29.578Z",
    "updated_at": "2018-01-21T12:33:07.419Z"
  },
```

The attribute `"data_option": {},` is causing the issue.

We were able to find the field by running the following line on the [Zammad rails console](https://docs.zammad.org/en/latest/admin-console.html#start-zammad-s-rails-console): `ObjectManager::Attribute.all.find { |a| a.data_option.blank? || !a.data_option.key?(:null) }`

Since in our case a `"data_type": "input"` field was affected we took the default value and updated the field via:

`ObjectManager::Attribute.find(1337).update!(data_option: {"type"=>"text", "maxlength"=>1000, "null"=>true})`

This might vary, depending on the field type of the affected attribute.

**Comment by rlue** [comment]
@veritaspoutine, have you had a chance to try @thorsteneckel's suggestion? To summarize:

* The migration that's failing for you was introduced in Zammad `stable-2.4`. That means you should be able to update to [`stable-2.3`](http://packager.io/gh/zammad/zammad/refs/stable-2.3) with no problems, and without having to incrementally upgrade from 2.1 to 2.2 to 2.3.

* The reason that the migration is failing is that there are some existing `ObjectManager::Attribute` records in your database that are missing a required attribute. Specifically, every `ObjectManager::Attribute` record should have a `data_option` that includes either `null: true` or `null: false`. As in Thorsten's example above, it's probably wise to add other relevant options, if they are missing.

  If you don't know what the other options should look like, tell us what the `data_type` attribute is, and we can help you decide.

* If you resolve this problem, you should be able to upgrade to 2.6 directly, without having to bother with upgrading to 2.3 first.

To get a complete list of all `ObjectManager::Attribute` records that are causing problems here, try the following command in the Rails console:

    ObjectManager::Attribute.select(:id, :data_type, :data_option).select { |a| [true, false].exclude?(a.data_option.try(:[], :null)) }

---

@thorsteneckel, since the problem is occurring in an existing migration, I think that if we wanted to prevent this problem in the future, we'd have to retroactively modify the broken migration, which is not a good practice AFAIU. 

Is there another way that I'm not seeing? Or do you think it might be better to treat this as a support issue, rather than a bug?

**Comment by SummerStorm** [comment]
@rlue  The failing migration seems to be [20180220171219_check_for_object_attributes.rb](https://github.com/zammad/zammad/blob/develop/db/migrate/20180220171219_check_for_object_attributes.rb)

I would suggest the following:

1. Commenting out the code in the existing migration so that it effectively does nothing:
```
def change
end
```
(Commenting it out is purely for documentation purposes. Deleting the lines would be fine too.)

2. Add a new migration one second behind the existing one: `20180220171220_check_for_object_attributes_fixed.rb`. 

The new migration is the correct rewrite of the existing one, with the added constraint that it must correctly skip valid attributes from all later versions of Zammad. 

This way DB migration will again work from any previous version to the latest version.

**Comment by rlue** [comment]
Thanks @SummerStorm, retroactively **inserting** a new migration does seem to work, although I wonder if it would just make more sense to place this logic in the model (say, in a `before_save` callback) instead. In fact, that might obviate the need for altering the migration history altogether.

In any case, this still leaves the question of what the default value of `data_option[:null]` should be. I don't know enough about the object manager and its use cases to say;

**Comment by veritaspoutine** [comment]
@thorsteneckel @rlue 
I would like to apologize for not getting back to you guys soon enough.  I tried @thorsteneckel's solution upgrading to 2.5 incrementally and it worked. I really appreciate all the helps from zammad and other contributors.

**Comment by thorsteneckel** [comment]
@veritaspoutine Thanks for your feedback! Especially the kind words 😊 Since there were some bigger changes involved and there is a practical workaround @rlue and I decided to fix this issue for Zammad 2.7. 🚀
